nvflare/apis/fl_component.py

```diff
@@ -228,3 +228,48 @@ def _fire_log_event(self, event_type: str, log_tag: str, log_msg: str, fl_ctx: F
         dxo = event_data.to_dxo()
         fl_ctx.set_prop(key=FLContextKey.EVENT_DATA, value=dxo.to_shareable(), private=True, sticky=False)
         self.fire_event(event_type=event_type, fl_ctx=fl_ctx)
+
+    def apply_data_filters(self, task_filter_list, task_data, fl_ctx):
+        filter_error = False
+        # apply scope filters first
+        scope_object = fl_ctx.get_prop(FLContextKey.SCOPE_OBJECT)
+        filter_list = []
+        if scope_object:
+            if scope_object.task_data_filters:
+                filter_list.extend(scope_object.task_data_filters)
+        if task_filter_list:
+            filter_list.extend(task_filter_list)
+        if filter_list:
+            for f in filter_list:
+                try:
+                    task_data = f.process(task_data, fl_ctx)
+                except Exception as e:
+                    self.log_exception(
+                        fl_ctx,
+                        "processing error in task data filter {}: {}; "
+                        "asked client to try again later".format(type(f), secure_format_exception(e)),
+                    )
+                    filter_error = True
+                    break
+        return filter_error, task_data
+
+    def apply_result_filters(self, task_filter_list, result, fl_ctx):
+        filter_error = False
+        filter_list = []
+        scope_object = fl_ctx.get_prop(FLContextKey.SCOPE_OBJECT)
+        if scope_object and scope_object.task_result_filters:
+            filter_list.extend(scope_object.task_result_filters)
+        if task_filter_list:
+            filter_list.extend(task_filter_list)
+        if filter_list:
+            for f in filter_list:
+                try:
+                    result = f.process(result, fl_ctx)
+                except Exception as e:
+                    self.log_exception(
+                        fl_ctx,
+                        "Error processing in task result filter {}: {}".format(type(f), secure_format_exception(e)),
```

Comment:
I notice in "task data filter"'s error message we have "asked client to try again later",
Do we need to have that here?

And if these lines are duplicated, maybe we can have have a common method and reuse it

Reply:
extracted the common _apply_filters()